Add API to control JS features available to content scripts

rniwa opened the pull request:
#### e42b814875a017b4a27630c14cf83774d2a96359
<pre>
Add API to control JS features available to content scripts
<a href="https://bugs.webkit.org/show_bug.cgi?id=282591">https://bugs.webkit.org/show_bug.cgi?id=282591</a>
<a href="https://rdar.apple.com/139704339">rdar://139704339</a>

Reviewed by Wenson Hsieh.

This PR introduces _WKContentWorldConfiguration, which controls whether content
script has the capability to autofill, access closed shadow roots, and disable
legacy behavior of element names overriding functions on document prototype.

For autofill, we expose 4 new attributes on HTMLInputElement:
 * autofillAvailable - boolean indicating whether autofill is available or not
   for users.
 * autofilled - boolean indicating whether the input element had been autofilled
   or not.
 * autofillVisibility - specifies the visibility of autofilled text. Possible
   values are &quot;normal&quot;, &quot;visible&quot;, and &quot;hidden&quot;.
 * autofillButtonType - specifies the type of autofill button to display.
   Possible values are &quot;none&quot;, &quot;credentials&quot;, &quot;contacts&quot;, &quot;strong-password&quot;,
   &quot;credit-card&quot;, and &quot;loading&quot;.

* Source/WebCore/bindings/js/DOMWrapperWorld.h:
(WebCore::DOMWrapperWorld::setAllowAutofill):
(WebCore::DOMWrapperWorld::allowAutofill const):
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::autofillVisibility const):
(WebCore::HTMLInputElement::setAutofillVisibility):
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebCore/html/HTMLInputElement.idl:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/PlatformMac.cmake:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h:
* Source/WebKit/Shared/ContentWorldData.h:
* Source/WebKit/Shared/ContentWorldData.serialization.in:
* Source/WebKit/Shared/ContentWorldShared.h:
* Source/WebKit/Shared/UserContentControllerParameters.h:
* Source/WebKit/Shared/UserContentControllerParameters.serialization.in:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/APIContentWorld.cpp:
(API::ContentWorld::ContentWorld):
(API::ContentWorld::sharedWorldWithName):
(API::ContentWorld::defaultClientWorld):
* Source/WebKit/UIProcess/API/APIContentWorld.h:
* Source/WebKit/UIProcess/API/Cocoa/WKContentWorld.mm:
(checkContentWorldOptions):
(+[WKContentWorld worldWithName:]):
(+[WKContentWorld _worldWithConfiguration:]):
* Source/WebKit/UIProcess/API/Cocoa/WKContentWorldConfiguration.mm: Added.
(-[_WKContentWorldConfiguration name]):
(-[_WKContentWorldConfiguration setName:]):
(-[_WKContentWorldConfiguration copyWithZone:]):
(+[_WKContentWorldConfiguration supportsSecureCoding]):
(-[_WKContentWorldConfiguration encodeWithCoder:]):
(-[_WKContentWorldConfiguration initWithCoder:]):
* Source/WebKit/UIProcess/API/Cocoa/WKContentWorldPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKContentWorldConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushSubscriptionData.mm:
* Source/WebKit/UIProcess/Cocoa/CoreTelephonyUtilities.mm:
* Source/WebKit/UIProcess/ios/PointerTouchCompatibilitySimulator.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundleScriptWorld.cpp:
(WebKit::InjectedBundleScriptWorld::setAllowAutofill):
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundleScriptWorld.h:
* Source/WebKit/WebProcess/UserContent/WebUserContentController.cpp:
(WebKit::WebUserContentController::addContentWorld):
(WebKit::WebUserContentController::addContentWorlds):
* Source/WebKit/WebProcess/UserContent/WebUserContentController.h:
* Source/WebKit/WebProcess/UserContent/WebUserContentController.messages.in:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::runJavaScriptInFrameInScriptWorld):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UserContentController.mm:
(TEST(WKUserContentController, AllowAutofill)):
(TEST(WKUserContentController, AllowAccessToClosedShadowRoots)):
(TEST(WKUserContentController, DisableLegacyBuiltinOverrides)):

Canonical link: <a href="https://commits.webkit.org/286778@main">https://commits.webkit.org/286778@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b40aae5eb8b2dffacf9033bddda67277b07e767c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77038 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56073 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29953 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81588 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28317 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79155 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65221 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4369 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60366 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18436 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80105 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50319 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66128 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40677 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/76544 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47721 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23626 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26643 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68838 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23954 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83022 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4418 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2965 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68651 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4573 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66101 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67904 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11878 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9965 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11922 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4364 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/7180 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4384 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7819 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6143 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/DerivedSources.make:
```diff
@@ -599,6 +599,7 @@ SERIALIZATION_DESCRIPTION_FILES = \
 	Shared/Cocoa/SharedCARingBuffer.serialization.in \
 	Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in \
 	Shared/CallbackID.serialization.in \
+    Shared/ContentWorldData.serialization.in \
```
Nit - looks like this uses spaces while the other lines use tabs.

rniwa replied:
Fixed.

whsieh commented on Source/WebKit/Shared/ContentWorldData.h:
```diff
@@ -0,0 +1,39 @@
+namespace WebKit {
+
+struct ContentWorldData {
+    WebKit::ContentWorldIdentifier identifier;
```
Nit - WebKit namespace is redundant here.

```suggestion
    ContentWorldIdentifier identifier;
```

rniwa replied:
Removed.

whsieh commented on Source/WebKit/WebProcess/UserContent/WebUserContentController.cpp:
```diff
@@ -104,28 +104,36 @@ InjectedBundleScriptWorld* WebUserContentController::worldForIdentifier(ContentW
-    if (addResult.isNewEntry)
-        return addResult.iterator->value.first.ptr();
+    if (addResult.isNewEntry) {
+        Ref scirptWorld = addResult.iterator->value.first;
```
```suggestion
        Ref scriptWorld = addResult.iterator->value.first;
```

rniwa replied:
Fixed.

pxlcoder commented on Source/WebKit/DerivedSources.make:
```diff
@@ -599,6 +599,7 @@ SERIALIZATION_DESCRIPTION_FILES = \
 	Shared/Cocoa/SharedCARingBuffer.serialization.in \
 	Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in \
 	Shared/CallbackID.serialization.in \
+    Shared/ContentWorldData.serialization.in \
```
Nit: Indentation.

rniwa replied:
Fixed.

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/WKContentWorldOptions.h:
```diff
@@ -0,0 +1,51 @@
+- If your scripts have to access autofill capabilities, you may want to set allowAutofill to YES. */
+WK_SWIFT_UI_ACTOR
+WK_CLASS_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA))
+@interface WKContentWorldOptions : NSObject
```
this should be annotated as Sendable, and should also be refined as `WKContentWorld.Options` for Swift

rniwa replied:
How do I do that?

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/WKContentWorldOptions.h:
```diff
@@ -0,0 +1,51 @@
+- If your scripts have to access autofill capabilities, you may want to set allowAutofill to YES. */
+WK_SWIFT_UI_ACTOR
+WK_CLASS_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA))
+@interface WKContentWorldOptions : NSObject
```
May as well make this conform to `NSSecureCoding`?

rniwa replied:
And NSCopying?

pxlcoder replied:
Yeah, I think `NSCopying` would also be good.

rniwa replied:
Done.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/UserContentController.mm:
```diff
@@ -1281,3 +1282,133 @@ - (void)userContentController:(WKUserContentController *)userContentController d
+    receivedScriptMessage = false;
+
+    auto options = adoptNS([[WKContentWorldOptions alloc] init]);
+    options.get().allowAutofill = YES;
```
```suggestion
    RetainPtr options = adoptNS([[WKContentWorldOptions alloc] init]);
    [options setAllowAutofill:YES];
```

rniwa replied:
Sure, changed.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/UserContentController.mm:
```diff
@@ -1281,3 +1282,133 @@ - (void)userContentController:(WKUserContentController *)userContentController d
+
+    RetainPtr<WKContentWorld> world = [WKContentWorld worldWithName:@"TestWorldAllowingAutofill" options:options.get()];
+    RetainPtr<ScriptMessageHandler> handler = adoptNS([[ScriptMessageHandler alloc] init]);
+    RetainPtr<WKUserScript> userScript = adoptNS([[WKUserScript alloc] _initWithSource:@"window.webkit.messageHandlers.testHandler.postMessage(typeof(document.createElement('input').autofillAvailable))"
```
Can remove the `<type>` for all of these. And instances below.

rniwa replied:
Good point. Removed.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/UserContentController.mm:
```diff
@@ -1281,3 +1282,133 @@ - (void)userContentController:(WKUserContentController *)userContentController d
+    NSURLRequest *request = [NSURLRequest requestWithURL:[NSBundle.test_resourcesBundle URLForResource:@"simple" withExtension:@"html"]];
+
+    isDoneWithNavigation = false;
+    [webView loadRequest:request];
```
```suggestion
    RetainPtr request = [NSURLRequest requestWithURL:[NSBundle.test_resourcesBundle URLForResource:@"simple" withExtension:@"html"]];

    isDoneWithNavigation = false;
    [webView loadRequest:request.get()];
```

rniwa replied:
Fixed/

pxlcoder reviewed (commented):
Test nits (same comments apply to the other tests).

whsieh commented on Source/WebKit/UIProcess/API/Cocoa/WKContentWorldOptions.mm:
```diff
@@ -0,0 +1,78 @@
+    if (!(self = [super init]))
+        return nil;
+    return self;
+}
```
Nit - I don't think this override is necessary.

whsieh commented on Source/WebKit/UIProcess/API/Cocoa/_WKContentWorldOptions.h:
```diff
@@ -0,0 +1,52 @@
+WK_CLASS_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA))
+@interface _WKContentWorldOptions : NSObject<NSCopying, NSSecureCoding>
+
+- (instancetype)init;
```
I _think_ you can omit this initializer.

rniwa committed:
Add API to control JS features available to content scripts
https://bugs.webkit.org/show_bug.cgi?id=282591
rdar://139704339

Reviewed by Wenson Hsieh.

This PR introduces _WKContentWorldConfiguration, which controls whether content
script has the capability to autofill, access closed shadow roots, and disable
legacy behavior of element names overriding functions on document prototype.

For autofill, we expose 4 new attributes on HTMLInputElement:
 * autofillAvailable - boolean indicating whether autofill is available or not
   for users.
 * autofilled - boolean indicating whether the input element had been autofilled
   or not.
 * autofillVisibility - specifies the visibility of autofilled text. Possible
   values are "normal", "visible", and "hidden".
 * autofillButtonType - specifies the type of autofill button to display.
   Possible values are "none", "credentials", "contacts", "strong-password",
   "credit-card", and "loading".

* Source/WebCore/bindings/js/DOMWrapperWorld.h:
(WebCore::DOMWrapperWorld::setAllowAutofill):
(WebCore::DOMWrapperWorld::allowAutofill const):
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::autofillVisibility const):
(WebCore::HTMLInputElement::setAutofillVisibility):
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebCore/html/HTMLInputElement.idl:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/PlatformMac.cmake:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/API/Cocoa/WebKitPrivate.h:
* Source/WebKit/Shared/ContentWorldData.h:
* Source/WebKit/Shared/ContentWorldData.serialization.in:
* Source/WebKit/Shared/ContentWorldShared.h:
* Source/WebKit/Shared/UserContentControllerParameters.h:
* Source/WebKit/Shared/UserContentControllerParameters.serialization.in:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/APIContentWorld.cpp:
(API::ContentWorld::ContentWorld):
(API::ContentWorld::sharedWorldWithName):
(API::ContentWorld::defaultClientWorld):
* Source/WebKit/UIProcess/API/APIContentWorld.h:
* Source/WebKit/UIProcess/API/Cocoa/WKContentWorld.mm:
(checkContentWorldOptions):
(+[WKContentWorld worldWithName:]):
(+[WKContentWorld _worldWithConfiguration:]):
* Source/WebKit/UIProcess/API/Cocoa/WKContentWorldConfiguration.mm: Added.
(-[_WKContentWorldConfiguration name]):
(-[_WKContentWorldConfiguration setName:]):
(-[_WKContentWorldConfiguration copyWithZone:]):
(+[_WKContentWorldConfiguration supportsSecureCoding]):
(-[_WKContentWorldConfiguration encodeWithCoder:]):
(-[_WKContentWorldConfiguration initWithCoder:]):
* Source/WebKit/UIProcess/API/Cocoa/WKContentWorldPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKContentWorldConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebPushSubscriptionData.mm:
* Source/WebKit/UIProcess/Cocoa/CoreTelephonyUtilities.mm:
* Source/WebKit/UIProcess/ios/PointerTouchCompatibilitySimulator.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundleScriptWorld.cpp:
(WebKit::InjectedBundleScriptWorld::setAllowAutofill):
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundleScriptWorld.h:
* Source/WebKit/WebProcess/UserContent/WebUserContentController.cpp:
(WebKit::WebUserContentController::addContentWorld):
(WebKit::WebUserContentController::addContentWorlds):
* Source/WebKit/WebProcess/UserContent/WebUserContentController.h:
* Source/WebKit/WebProcess/UserContent/WebUserContentController.messages.in:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::runJavaScriptInFrameInScriptWorld):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UserContentController.mm:
(TEST(WKUserContentController, AllowAutofill)):
(TEST(WKUserContentController, AllowAccessToClosedShadowRoots)):
(TEST(WKUserContentController, DisableLegacyBuiltinOverrides)):

Canonical link: https://commits.webkit.org/286778@main
